python/hidet/graph/ops/definitions/reduce/reduce.py

```diff
@@ -70,23 +69,161 @@ def reduce_fcompute(*reduce_indices):
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 16, 32))
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 8, 32))
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 4, 32))
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 2, 32))
+                rv = ro.combine(rv, shfl_down_sync(mask, rv, 1, 32))
+                rv = shfl_sync(mask, rv, 0, 32)
+                rv = ro.finalize(acc=rv, size=reduce_extent)
+
+                if threadIdx.x == 0:
+                    for indices in remain_layout.on(blockIdx.x):
+                        y.write(indices, cast(rv, xdtype), protected=False)
+
+        ir_module = module.ir_module()
+        return ir_module
+
+    def cuda_schedule_reduce_by_default(self) -> IRModule:
+        import hidet
+        from hidet.ir.compute import ReduceOperation
+        from hidet.ir.type import data_type
+        from hidet.lang import spatial, repeat, attrs
+        from hidet.lang.cuda import blockIdx, threadIdx, register_tensor
+
+        x, y = self.inputs[0], self.outputs[0]
+        dims = self.dims
+        shape: List[int] = list(x.const_shape)
+        xdtype = self.xdtype
+
+        if self.keep_dim:
+            remain_shape = [v if i not in dims else 1 for i, v in enumerate(shape)]
+        else:
+            remain_shape = [v for i, v in enumerate(shape) if i not in dims]
+
+        remain_extent = hidet.utils.prod(remain_shape)
+        reduce_shape = [shape[i] for i in dims]
+        reduce_extent = hidet.utils.prod(reduce_shape)
+        block_size = 256 if 256 < remain_extent else remain_extent
+        remain_layout = spatial(*remain_shape)
+
+        spatial_shape = []
+        repeat_shape = []
+        for i in range(len(shape)):
+            if i in dims:
+                spatial_shape.append(1)
+                repeat_shape.append(shape[i])
+            else:
+                spatial_shape.append(shape[i])
+                repeat_shape.append(1)
+        task_layout = repeat(*repeat_shape) * spatial(*spatial_shape)
+
+        grid_size = (remain_layout.num_workers + block_size - 1) // block_size
+        accumulate_dtype = self.attrs['accumulate_dtype']
+        reduce_type = self.attrs['reduce_type']
+        ro = ReduceOperation.from_name(reduce_type)
+
+        with hidet.script_module() as module:
+
+            @hidet.script
+            def reduce_kernel(x: xdtype[x.const_shape], y: xdtype[y.const_shape]):
+                # Each 256-thread ThreadBlock handles 512 columns
+                attrs.cuda.grid_dim = grid_size
+                attrs.cuda.block_dim = block_size
+                attrs.cuda.min_blocks = 1
+
+                rv = register_tensor(accumulate_dtype, [1])
+                rv[0] = ro.initial_value(data_type(accumulate_dtype))
+
+                if threadIdx.x + blockIdx.x * block_size < remain_extent:
+                    for indices in task_layout.on(threadIdx.x + blockIdx.x * block_size):
```

Comment:
I prefer to make the task mapping definition in the script function when it is only used once, which is more readable for me.